### PR TITLE
Allow custom hooks from config section

### PR DESCRIPTION
**Pull request by shaffe-fr**
Hi,
This prevent adding test command for non valid hooks.
For now every scripts of the file and the config section are listed in the possible commands.
Thanks

### CODE REVIEW

**Comment by BrainMaestro** [comment]
@shaffe-fr this is an intentional change. see https://github.com/BrainMaestro/composer-git-hooks/pull/111

**Comment by shaffe-fr** [comment]
Okay @BrainMaestro, it was funny to see stuff from the scripts section.

**Comment by BrainMaestro** [comment]
hmm actually they should only be visible with the `--allow-custom-hooks` option. how do you think this should be handled? @micc83 any ideas?

**Comment by shaffe-fr** [comment]
Maybe only show hooks from `extra.hooks`, without `config`?

**Comment by BrainMaestro** [comment]
yeah I like that

**Comment by shaffe-fr** [comment]
This can be a BC for projects using `--allow-custom-hooks`. A v2 might be needed.

**Comment by micc83** [comment]
@BrainMaestro this was indeed one of my concerns, stated in https://github.com/BrainMaestro/composer-git-hooks/issues/109#issue-688188440. Unfortunately I'm not sure how to handle it without digging further.

**Comment by shaffe-fr** [comment]
What do you think about enabling at a project level with in a new config option @micc83?
The `--allow-custom-flags` would be removed.

**Comment by micc83** [comment]
@shaffe-fr Well, from an usability point of view I guess it would be even better. 

Also I see that `extra.hooks.config` has been used before to specify package specific configs and avoid cluttering the main namespace.
```json
{
    "extra": {
        "hooks": {
            "config": {
                "stop-on-failure": ["pre-push"],
                "allow-custom-hooks": true
            },
            "pre-push": [
                "php-cs-fixer fix --dry-run --stop-on-violation .",
                "phpunit"
            ],
        }
    }
}
```

**Comment by shaffe-fr** [comment]
I'll update this PR so Hook::getValidHooks use this new config.

**Comment by micc83** [comment]
@shaffe-fr I would wait for some feedback from @BrainMaestro.

**Comment by shaffe-fr** [comment]
I think he is ok https://github.com/BrainMaestro/composer-git-hooks/pull/117#issuecomment-783241902.

**Comment by BrainMaestro** [comment]
@shaffe-fr @micc83 I like the suggestion. Especially the second one that is an array instead of just a boolean flag.

**Comment by shaffe-fr** [comment]
Ok @BrainMaestro, I'll work on that :)

**Comment by shaffe-fr** [comment]
Hi @BrainMaestro @micc83,
Here is the rework of custom hooks. 
As mentionned above, this is a breaking change and should be the first release of version 2 :)

File: src/Hook.php [hunk 1]
```diff
@@ -134,4 +134,41 @@ private static function getDefaultHooks()
...
+
+        if (! is_array($customHooks)) {
+            throw new Exception('Custom hooks must be an array.');
+        }
+
+        $configIndex = array_search('config', $customHooks);
+
+        if ($configIndex !== false) {
+            unset($customHooks[$configIndex]);
+        }
```
Comment (BrainMaestro) [inline src/Hook.php:170]: I'm not sure I understand why we have this

Comment (shaffe-fr) [reply src/Hook.php:170]: I think that `config` should not be used as a hook. You're thought on this?

File: src/Hook.php [hunk 2]
```diff
@@ -72,14 +72,14 @@ public static function stopHookOnFailure($dir, $hook)
     /**
      * Get scripts section of the composer config file.
      *
-     * @param	$dir	string	dir where to look for composer.json
+     * @param  string $dir Directory where to look for composer.json
      *
      * @return array
      */
     public static function getValidHooks($dir)
```
[the remark this answers is not in the document]
Comment (shaffe-fr) [reply src/Hook.php:79]: I'll check that.

File: src/Hook.php (hunk 1 above)
Comment (BrainMaestro) [reply src/Hook.php:170]: Oh that's right! Makes sense. Could you make a note of this in the readme section for custom hooks that `config` is not allowed

File: src/Hook.php (hunk 2 above)
[the remark this answers is not in the document]
Comment (shaffe-fr) [reply src/Hook.php:74]: I pushed a new version with the two methods combined.
As this PR is breaking, I cleaned the undocumented legacy behaviour to fetch hooks from the `scripts` and top level `hooks` sections.

[the remark this answers is not in the document]
Comment (BrainMaestro) [reply src/Hook.php:74]: alright, if we're removing the old support, then we need to adjust the section of the docs that says 
```The previous way of adding hooks to the `scripts` section of your `composer.json` is still supported, but this way is cleaner if you have many scripts```

[the remark this answers is not in the document]
Comment (shaffe-fr) [reply src/Hook.php:74]: Sorry, I didn't saw that sentence. I moved the removal to a new commit.

File: README.md [hunk 3]
```diff
@@ -40,6 +40,8 @@ composer require --dev brainmaestro/composer-git-hooks
 
 This installs the `cghooks` binary to your `vendor/bin` folder. If this folder is not in your path, you will need to preface every command with `vendor/bin/`.
 
+Note: hooks declared in the `scripts` or `hooks` root sections of `composer.json` are no longer supported in v2.
```
Comment (BrainMaestro) [inline README.md:43]: v3. we're already in v2

**Review by BrainMaestro — APPROVED** [review]

**Comment by BrainMaestro** [comment]
@shaffe-fr thanks again for the great contribution!

I will give this a bit of time (roughly a week) before I make a new release

**Comment by shaffe-fr** [comment]
It was my pleasure.
Good idea to wait a bit, I'm not sure how many people will be affected by the BC.
Do you have a some thoughts about that?

**Pull request merged**